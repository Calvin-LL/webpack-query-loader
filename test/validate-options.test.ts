import webpack from "webpack";

import WQLWebpackTestCompiler from "./helpers/WQLWebpackTestCompiler";

describe.each([4, 5] as const)("v%d validate options", (webpackVersion) => {
  const tests = {
    use: {
      success: [
        "file-loader",
        { loader: "file-loader" },
        { loader: "file-loader", options: { name: "[path][name].[ext]" } },
      ],
      failure: [{ options: { name: "[path][name].[ext]" } }, true],
    },
    resourceQuery: {
      success: ["test", ["test", "!test1"], () => true],
      failure: [0],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ): void {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = new WQLWebpackTestCompiler({ webpackVersion });

      let stats: webpack.Stats | undefined;

      try {
        stats = (
          await compiler.compile({
            loaderOptions: {
              use: "file-loader",
              [key]: value,
            },
            throwOnError: false,
          })
        ).stats;
      } finally {
        if (type === "success") {
          expect(stats!.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const errors = stats!.compilation.errors;

          expect(errors).toHaveLength(1);
          expect(errors).toMatchSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
