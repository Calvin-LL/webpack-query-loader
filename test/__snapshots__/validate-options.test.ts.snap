// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v4 validate options should throw an error on the "resourceQuery" option with 0 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from <rootDir>/test-dist/cjs.js):
ValidationError: Invalid options object. Query Loader has been initialized using an options object that does not match the API schema.
 - options.resourceQuery should be one of these:
   string | [any, ...] | function
   -> A condition that needs to be satisfied before the loader runs
   Details:
    * options.resourceQuery should be a string.
    * options.resourceQuery should be an array:
      [any, ...]
    * options.resourceQuery should be an instance of function.
    at Object.validate (<rootDir>/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.pitch (<rootDir>/src/index.ts:53:3)],
]
`;

exports[`v4 validate options should throw an error on the "use" option with {"options":{"name":"[path][name].[ext]"}} value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from <rootDir>/test-dist/cjs.js):
Error: Loader Not Found
    at normalizeUse (<rootDir>/src/index.ts:204:41)
    at Object.pitch (<rootDir>/src/index.ts:69:64)],
]
`;

exports[`v4 validate options should throw an error on the "use" option with true value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from <rootDir>/test-dist/cjs.js):
ValidationError: Invalid options object. Query Loader has been initialized using an options object that does not match the API schema.
 - options.use should be one of these:
   string | object { loader?, options?, … }
   -> Loader
   Details:
    * options.use should be a string.
    * options.use should be an object:
      object { loader?, options?, … }
    at Object.validate (<rootDir>/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.pitch (<rootDir>/src/index.ts:53:3)],
]
`;

exports[`v5 validate options should throw an error on the "resourceQuery" option with 0 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Query Loader has been initialized using an options object that does not match the API schema.
 - options.resourceQuery should be one of these:
   string | [any, ...] | function
   -> A condition that needs to be satisfied before the loader runs
   Details:
    * options.resourceQuery should be a string.
    * options.resourceQuery should be an array:
      [any, ...]
    * options.resourceQuery should be an instance of function.
    at Object.validate (<rootDir>/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.pitch (<rootDir>/src/index.ts:53:3)],
]
`;

exports[`v5 validate options should throw an error on the "use" option with {"options":{"name":"[path][name].[ext]"}} value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
Error: Loader Not Found
    at normalizeUse (<rootDir>/src/index.ts:204:41)
    at Object.pitch (<rootDir>/src/index.ts:69:64)],
]
`;

exports[`v5 validate options should throw an error on the "use" option with true value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Query Loader has been initialized using an options object that does not match the API schema.
 - options.use should be one of these:
   string | object { loader?, options?, … }
   -> Loader
   Details:
    * options.use should be a string.
    * options.use should be an object:
      object { loader?, options?, … }
    at Object.validate (<rootDir>/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.pitch (<rootDir>/src/index.ts:53:3)],
]
`;
